### PR TITLE
[DCP] Adds support for meta tensor loading for DCP.load_state_dict()

### PR DESCRIPTION
**Pull request by wz337**
Currently, DCP requires the `model.state_dict()` to be materialized before passing it to DCP to load, since DCP uses the pre-allocated storage from the initialized model state_dict. Therefore, even for fine-tuning and distributed inference, users would need to explicitly materialize the model on GPU before `DCP.load_state_dict()`. 

Today's flow:
```
with torch.device("meta"):
    model2 = parallelize_module(
        MLPModule("meta"), tp_mesh, parallelize_plan=parallelize_plan
    )

model.to_empty(device='cuda')
state_dict_to_load = model2.state_dict()
DCP.load_state_dict(
    state_dict=state_dict_to_load,
    storage_reader=DCP.FileSystemReader(CHECKPOINT_DIR),
)
model2.load_state_dict(state_dict_to_load)
```

This PR adds support for meta tensor loading. In DCP's planner, when encountering tensors/DTensor on meta device, we initialize tensor/DTensor on the current device on the fly and replace the tensor/DTensor on meta device in the state_dict.  After the change, users no longer needs to manually call `model.to_empty()` when loading existing checkpoints for fine-tuning and distributed inference.

Updated user flow:
```
with torch.device("meta"):
    model2 = parallelize_module(
        MLPModule("meta"), tp_mesh, parallelize_plan=parallelize_plan
    )
# no longer need to call model.to_empty(device='cuda')
state_dict_to_load = model2.state_dict()
DCP.load_state_dict(
    state_dict=state_dict_to_load,
    storage_reader=DCP.FileSystemReader(CHECKPOINT_DIR),
)
model2.load_state_dict(state_dict_to_load, assign=True)
```

Note that for distributed training, it's still the users' responsibility to reset the parameters (`model.reset_parameters()`) as checkpoint might not exist.

Note that we need to loop thru the state_dict to replace meta tensor/DTensor instead of calling `model.to_empty()` since `DCP.load()` only takes in state_dict but not model. 

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @tianyu-l @wconstab @yf225 @LucasLLC @kiukchung @d4l3k @lucasllc

### CODE REVIEW

File: torch/distributed/checkpoint/default_planner.py [hunk 1]
```diff
@@ -178,6 +179,7 @@ def set_up_planner(
         metadata: Metadata,
         is_coordinator: bool,
     ) -> None:
+        state_dict = _initialize_meta_tensors(state_dict)
```
Comment (fegin) [inline torch/distributed/checkpoint/default_planner.py:182]: This won't work. What if the state_dict is nested?

You can write something like
```
def init_meta(v):
      ...
      if isinstance(value, DTensor):
           ...
      elif isinstance(value, ShardedTensor):
           ...
      else:
           return torch.empty_like(value, device=device)
  
  state_dict = tree_map_only(torch.Tensor, init_meta)        
 ```

**Review by fegin — CHANGES_REQUESTED** [review]

File: test/distributed/checkpoint/test_tp_checkpoint.py [hunk 2]
```diff
@@ -66,16 +79,57 @@ def test_tp_checkpoint(self):
...
+        model = UnevenShardedModel(self.device_type).cuda(self.rank)
+        # Parallelize the module based on the given Parallel Style.
+        parallelize_plan = {
+            "net1": RowwiseParallel(),
+            "net2": ColwiseParallel(),
+        }
+        model = parallelize_module(model, tp_mesh, parallelize_plan=parallelize_plan)
+        original_state_dict = deepcopy(model.state_dict())
+
+        DCP.save_state_dict(
```
Comment (LucasLLC) [inline test/distributed/checkpoint/test_tp_checkpoint.py:110]: Might be a good opportunity to use the new Checkpointer class

File: torch/distributed/checkpoint/planner_helpers.py [hunk 3]
```diff
@@ -253,3 +258,46 @@ def _create_read_items(fqn: str, md: STORAGE_TYPES, obj: Any) -> List[ReadItem]:
...
+
+
+def _init_meta_tensor(value: Any) -> Any:
+    """
+    Initializes tensor, moves it to device for torch.Tensor/DTensor on meta device.
+    """
+
+    device = getattr(value, "device", None)
+    # DCP does the initialization if it's meta tensor/DTensor.
+    if device and device == torch.device("meta"):
```
Comment (LucasLLC) [inline torch/distributed/checkpoint/planner_helpers.py:280]: Curious if the first `device` is necessary ?

**Review by LucasLLC — APPROVED** [review]
LGTM, just some small questions

**Review by fegin — APPROVED** [review]

File: test/distributed/checkpoint/test_tp_checkpoint.py (hunk 2 above)
Comment (LucasLLC) [reply test/distributed/checkpoint/test_tp_checkpoint.py:110]: Feel free to disregard since we are making Checkpointer private

Comment (wz337) [reply test/distributed/checkpoint/test_tp_checkpoint.py:110]: Ye. This was created back in November so I was still using our legacy setup. Can update it later.

File: torch/distributed/checkpoint/planner_helpers.py (hunk 3 above)
Comment (wz337) [reply torch/distributed/checkpoint/planner_helpers.py:280]: Good catch. This is redundant.

**Pull request closed without merging**